server/src/test/java/org/elasticsearch/cluster/service/MasterServiceTests.java

```diff
@@ -505,33 +520,31 @@ public String toString() {
             }
         }
 
-        int numberOfThreads = randomIntBetween(2, 8);
-        int taskSubmissionsPerThread = randomIntBetween(1, 64);
-        int numberOfExecutors = Math.max(1, numberOfThreads / 4);
+        final int numberOfThreads = randomIntBetween(2, 8);
+        final int taskSubmissionsPerThread = randomIntBetween(1, 64);
+        final int numberOfExecutors = Math.max(1, numberOfThreads / 4);
         final Semaphore semaphore = new Semaphore(numberOfExecutors);
```

Comment:
This semaphore seems unnecessary to me, maybe I'm missing something?

Reply:
If `ClusterStateTaskListener#clusterStateProcessed` is called after the `ClusterStateTaskExecutor#clusterStatePublished` then we can guarantee all tasks are completed just with `processedStatesLatch`. Let me confirm this assumption

Reply:
I was missing something: I didn't see the final call to `Semaphore#acquire` 🤦

Reply:
It looks like the order of execution is: 
* `ClusterStateTaskExecutor#execute`
* N * `ClusterStateTaskListener#clusterStateProcessed`
* `ClusterStateTaskExecutor#clusterStatePublished(ClusterStatePublicationEvent clusterPublicationEvent)`

This guards us from failing following assertion: `assertEquals(executor.batches.get(), executor.published.get());`

Reply:
As discussed in another channel, there's only one thread here so this can be a `new Semaphore(1)` instead, and the `semaphore.acquire()` call that happens when the task is executed should be `assertTrue(semaphore.tryAcquire())` instead.